pkg/runctl/controller.go

```diff
@@ -412,11 +410,47 @@ func (c *Controller) syncHandler(key string) error {
 		if err = c.changeAndCommitStateAndMeter(ctx, pipelineRun, api.StateWaiting, metav1.Now()); err != nil {
 			return err
 		}
+
 	case api.StateWaiting:
 		run, err := runManager.GetRun(ctx, pipelineRun)
 		if err != nil {
 			return c.onGetRunError(ctx, pipelineRunAPIObj, pipelineRun, err, api.StateCleaning, api.ResultErrorInfra, "waiting failed")
 		}
+		// the configuration should be loaded once per sync to avoid inconsistencies
+		// in case of concurrent configuration changes
+		pipelineRunsConfig, err := c.loadPipelineRunsConfig(ctx)
+		if err != nil {
+			return c.onGetRunError(ctx, pipelineRunAPIObj, pipelineRun, err, api.StateCleaning, api.ResultErrorInfra, "failed to load configuration for pipeline runs")
+		}
+
+		// Check for wait timeout
+		startTime := pipelineRun.GetStatus().StateDetails.StartedAt
+		timeout := pipelineRunsConfig.TimeoutWait
+		if startTime.Add(timeout.Duration).Before(time.Now()) {
+			return c.handleResultError(ctx, pipelineRun, api.ResultErrorInfra, "waiting failed", fmt.Errorf("wait timeout reached"))
+		}
+
+		if run == nil {
```

Comment:
If I'm not wrong `run` can never be nil here.
`run` returned by `runManager.GetRun` in line 415 can only be nil if also an `err` is returned. And in this case we will directly `return c.onGetRunError` in line 417.

Reply:
Or, if Tekton would return a nil run and nil err [here](https://github.com/SAP/stewardci-core/blob/master/pkg/runctl/run_manager.go#L670-L684), we would still wrap the nil TaskRun into a non-nil [`tektonRun`](https://github.com/SAP/stewardci-core/blob/master/pkg/runctl/run.go#L14-L21) struct.

Should [`GetRun`](https://github.com/SAP/stewardci-core/blob/master/pkg/runctl/run_manager.go#L671) return `nil, nil` in case the TaskRun has not been "started" yet? Then `run would be nil here.

Reply:
If task run cannot be found nil,nil is returned. See https://github.com/SAP/stewardci-core/blob/PR_tekton_retry/pkg/runctl/run_manager.go#L713.
This was changed in this PR.

Reply:
ah, ok